### PR TITLE
Added copy boundary to manifold id function.

### PR DESCRIPTION
**Pull request by luca-heltai**
This addresses in part #410, following @tjhei and @bangerth suggestions.

### CODE REVIEW

File: include/deal.II/grid/grid_tools.h [hunk 1]
```diff
@@ -1145,6 +1145,40 @@ namespace GridTools
...
+  /*@{*/
+
+  /**
+   * Copy boundary ids to manifold ids. The default manifold_id for
+   * new Triangulation objects is numbers::invalid_manifold_id. When
+   * refinements occurs, the Triangulation asks where to locate new
+   * points to the underlying manifold, and if the manifold_id of a
+   * boundary face is set to numbers::invalid_manifold_id, then
+   * Triangulation reverts back boundary_id (this was the behavior of
+   * deal.II < 8.2). This function copies the boundary_ids of the the
```
Comment (tjhei) [inline include/deal.II/grid/grid_tools.h]: remove "the"

File: include/deal.II/grid/grid_tools.h [hunk 2]
```diff
@@ -1145,6 +1145,40 @@ namespace GridTools
...
+   * boundary faces to the manifold_ids of the same faces, allowing
+   * the user to change the boundary_ids and use them for boundary
+   * conditions regardless of the geometry, which will use
+   * manifold_ids to create new points. Only active cells will be
+   * iterated over. This is a function you'd tipically call when there
```
Comment (tjhei) [inline include/deal.II/grid/grid_tools.h]: "typically"

**Comment by tjhei** [comment]
Looks good otherwise.

**Comment by luca-heltai** [comment]
My "tipical" mistake... :-). Fixing it.

**Comment by luca-heltai** [comment]
Wait before merging. I'm also adding the function which copies over the material_ids.

**Comment by luca-heltai** [comment]
With this, we have a way to copy both material and boundary ids over to manifold ids. 
Let me know if you want me to change anything.

File: include/deal.II/grid/grid_tools.h [hunk 3]
```diff
@@ -1145,6 +1145,77 @@ namespace GridTools
...
+  /*@{*/
+
+  /**
+   * Copy boundary ids to manifold ids. The default manifold_id for
+   * new Triangulation objects is numbers::invalid_manifold_id. When
+   * refinements occurs, the Triangulation asks where to locate new
+   * points to the underlying manifold, and if the manifold_id of a
+   * boundary face is set to numbers::invalid_manifold_id, then
+   * Triangulation reverts back boundary_id (this was the behavior of
+   * deal.II < 8.2). This function copies the boundary_ids of the
```
Comment (bangerth) [inline include/deal.II/grid/grid_tools.h]: I would remove this discussion of behavior. It is documented elsewhere and if we ever change it this comment will become stale. I would restrict the documentation to what the function does and, if appropriate to a rationale.

File: include/deal.II/grid/grid_tools.h [hunk 4]
```diff
@@ -1145,6 +1145,77 @@ namespace GridTools
...
+   * manifold_id. By default, boundary_ids are left untouched.
+   *
+   * @ingroup manifold
+   * @ingroup boundary
+   *
+   * @author Luca Heltai, 2015
+   */
+  template <int dim, int spacedim>
+  void copy_boundary_to_manifold_id(Triangulation<dim, spacedim> &tria,
+                                    bool reset_boundary_ids=false);
```
Comment (bangerth) [inline include/deal.II/grid/grid_tools.h]: Make the argument const

File: include/deal.II/grid/grid_tools.h [hunk 5]
```diff
@@ -1145,6 +1145,77 @@ namespace GridTools
...
+
+  /**
+   * Copy material ids to manifold ids. The default manifold_id for
+   * new Triangulation objects is numbers::invalid_manifold_id. When
+   * refinements occurs, the Triangulation asks where to locate new
+   * points to the underlying manifold.
+   *
+   * When reading a Triangulation from a supported input format,
+   * tipical informations that can be stored in a file are boundary
```
Comment (bangerth) [inline include/deal.II/grid/grid_tools.h]: Typo

File: include/deal.II/grid/grid_tools.h [hunk 6]
```diff
@@ -1145,6 +1145,77 @@ namespace GridTools
...
+   * conditions (which we store in boundary_ids), material types
+   * (which we store in material_ids) and in some cases subdomain ids
+   * (which we store in subdomain_ids).
```
Comment (bangerth) [inline include/deal.II/grid/grid_tools.h]: I would add that we store boundary ids for faces and edges, and material ids for cells

File: source/grid/grid_tools.cc [hunk 7]
```diff
@@ -2894,6 +2894,49 @@ namespace GridTools
...
+        cell->set_manifold_id(cell->material_id());
+        if (compute_face_ids == true)
+          {
+            for (unsigned int f=0; f<GeometryInfo<dim>::faces_per_cell; ++f)
+              {
+                cell->face(f)->set_manifold_id(cell->material_id());
+                if (cell->neighbor(f) != endc)
+                  cell->face(f)->set_manifold_id
+                  (std::min(cell->material_id(),
+                            cell->neighbor(f)->material_id()));
```
Comment (bangerth) [inline source/grid/grid_tools.cc]: Put the first set_manifold_id call into an else branch of the if. This way you have to call it only once if there is a neighbor.

File: include/deal.II/grid/grid_tools.h [hunk 8]
```diff
@@ -1145,6 +1145,77 @@ namespace GridTools
...
+   * selected for each face of the Triangulation. By default, face
+   * manifold_ids are not computed.
+   *
+   * @ingroup manifold
+   *
+   * @author Luca Heltai, 2015
+   */
+  template <int dim, int spacedim>
+  void copy_material_to_manifold_id(Triangulation<dim, spacedim> &tria,
+                                    bool compute_face_ids=false);
```
Comment (bangerth) [inline include/deal.II/grid/grid_tools.h]: Same here: make const

**Comment by bangerth** [comment]
Please merge after addressing these minor points.

**Pull request merged**